### PR TITLE
Various fixes/improvements related to registration & environments

### PR DESCRIPTION
**Pull request by ptoscano**
In case an activation key is used to register to a Candlepin with 'environments' resources, do not try to check/query for environments: the activation key carries the information on environments already.

To do the case easily, refactor a bit `RegisterCommand._process_environments()`:
- drop the `options` parameter, as it is the same as the `options` class variable that is already used inside `_process_environments()`
- factorize the behaviour for the "not supports_environments" case:
  - still exit with an error if --environments was specified
  - explicitly return early from the function with `None` otherwise, instead of doing it implicitly because of the lack of return in the "if supports_environments" that used to follow
- remove the "if supports_environments" check, as it is no more needed since the opposite is checked earlier on

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2046516
Card ID: ENT-4679

### CODE REVIEW

File: src/subscription_manager/cli_command/register.py [hunk 1]
```diff
@@ -329,39 +327,45 @@ def _prompt_for_environment(self):
...
+                return all_env_list[0]['id']
+
+            env_name_list = [env['name'] for env in all_env_list]
+            print(_('Hint: Organization "{key}" contains following environments: {list}').format(
+                  key=owner_key, list=", ".join(env_name_list)))
+            environments = self._prompt_for_environment()
+            if not self.cp.has_capability(MULTI_ENV) and ',' in environments:
+                system_exit(os.EX_USAGE, _("The entitlement server does not allow multiple environments"))
+
+        return check_set_environment_names(all_env_list, environments)
```
Comment (cnsnyder) [inline src/subscription_manager/cli_command/register.py:368]: all_env_list can be referenced before assignment if at least one environment is specified (along with the other necessary parameters for registration):
```subscription-manager register --serverurl SOME_SATELLITE/rhsm --environments THING --activationkey ABCD --org default --insecure```

```
local variable 'all_env_list' referenced before assignment
```

I'd suggest that check for duplicates separately from checking for environments which don't exist on the server. IMO we should only get the list of all environments when one is needed but none are specified

**Review by cnsnyder — CHANGES_REQUESTED** [review]
Please see inline comments :)
Thanks for the PR!

File: src/subscription_manager/cli_command/register.py (hunk 1 above)
Comment (ptoscano) [reply src/subscription_manager/cli_command/register.py:368]: It isn't just removing duplicates, it is also turning environment names into ids :-/

**Comment by ptoscano** [comment]
I dropped the optimization (which wasn't possible, like @cnsnyder mentioned, and I foolish missed), and also the enablement of `--environments` with `--activationkey`, as it poses additional issues currently.

File: src/subscription_manager/cli_command/register.py [hunk 2]
```diff
@@ -329,39 +329,45 @@ def _prompt_for_environment(self):
...
+
+        all_env_list = admin_cp.getEnvironmentList(owner_key)
+        if self.options.environments:
+            environments = self.options.environments
+        else:
+            # If there aren't any environments, don't prompt for one
+            if not all_env_list:
+                return None
+
+            # If the envronment list is len 1, pick that environment
```
Comment (jirihnidek) [inline src/subscription_manager/cli_command/register.py:358]: Please fix typo 'envronment'.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Overall it looks good to me. I have only small request.

BTW: shouldn't we add at least one unit test for the case? I mean the `_process_environments()` should return None, when activation key is used for registration.

File: src/subscription_manager/cli_command/register.py [hunk 3]
```diff
@@ -329,39 +329,45 @@ def _prompt_for_environment(self):
...
-                      key=owner_key, list=", ".join(env_name_list)))
-                environments = self._prompt_for_environment()
-                if not self.cp.has_capability(MULTI_ENV) and ',' in environments:
-                    system_exit(os.EX_USAGE, _("The entitlement server does not allow multiple environments"))
-
-            return check_set_environment_names(all_env_list, environments)
+        # We have an activation key, so don't need to fill/check the bits
+        # related to environments, as they are part of the activation key
+        if self.options.activation_keys:
+            return None
```
Comment (cnsnyder) [inline src/subscription_manager/cli_command/register.py:348]: Don't we want to allow users to set additional environments along with those on any activation keys in use?

Comment (jirihnidek) [reply src/subscription_manager/cli_command/register.py:348]: I'm not sure, but current implementation do not allow to use `--environments`, when `--activation-key` is used... BTW: it is possible to use --activation-key multiple time (with different environments) and it is possible to modify environments later using `subscription-manager environments --set`

File: src/subscription_manager/cli_command/register.py (hunk 2 above)
Comment (ptoscano) [reply src/subscription_manager/cli_command/register.py:358]: it is: it is not related to my changes

File: src/subscription_manager/cli_command/register.py (hunk 3 above)
Comment (cnsnyder) [reply src/subscription_manager/cli_command/register.py:348]: In that case I have no objection to this.

File: src/subscription_manager/cli_command/register.py (hunk 2 above)
Comment (jirihnidek) [reply src/subscription_manager/cli_command/register.py:358]: What about adding another commit to this PR?

Comment (ptoscano) [reply src/subscription_manager/cli_command/register.py:358]: I'd rather introduce tools like codespell for the whole codebase and fix all the typos/issues at once, rather than ad-hoc approach for every small typo...

Comment (jirihnidek) [reply src/subscription_manager/cli_command/register.py:358]: I'm 100% for fixing all typos. Great idea! Automated check of typos as CI also good idea.

Can you just fix this typo, when I asked for that and I said please? Thanks.

Comment (ptoscano) [reply src/subscription_manager/cli_command/register.py:358]: It's out of scope of this PR, so no. Thanks for asking though.

**Comment by jirihnidek** [comment]
Sounds good.
